.gitlab-ci.yml

```diff
@@ -0,0 +1,107 @@
+
+stages:
+  - lint
+  - test
+  - build
+
+variables:                         &default-vars
+  GIT_STRATEGY:                    fetch
+  GIT_DEPTH:                       100
+  CARGO_INCREMENTAL:               0
+  CARGO_TARGET_DIR:                "/ci-cache/${CI_PROJECT_NAME}/targets/${CI_COMMIT_REF_NAME}/${CI_JOB_NAME}"
+  CI_IMAGE:                        "paritytech/ci-linux:production"
+
+.rust-info-script:                 &rust-info-script
+  - rustup show
+  - cargo --version
+  - rustup +nightly show
+  - cargo +nightly --version
+  - bash --version
+  - sccache -s
+
+.docker-env:                       &docker-env
+  image:                           "${CI_IMAGE}"
+  before_script:
+    - *rust-info-script
+    - ./scripts/ci/pre_cache.sh
+    - sccache -s
+  retry:
+    max: 2
+    when:
+      - runner_system_failure
+      - unknown_failure
+      - api_failure
+  interruptible:                   true
+  tags:
+    - linux-docker
+
+.kubernetes-env:                   &kubernetes-env
+  retry:
+    max: 2
+    when:
+      - runner_system_failure
+      - unknown_failure
+      - api_failure
+  interruptible:                   true
+  tags:
+    - kubernetes-parity-build
+
+.collect-artifacts:                &collect-artifacts
+  artifacts:
+    name:                          "${CI_JOB_NAME}_${CI_COMMIT_REF_NAME}"
+    when:                          on_success
+    expire_in:                     28 days
+    paths:
+      - ./artifacts/
+
+#### stage:                        lint
+
+fmt:
+  stage:                           lint
+  image:                           "${CI_IMAGE}"
+  <<:                              *kubernetes-env
+  script:
+    - cargo +nightly fmt --all -- --check --config merge_imports=true
+
+clippy:
+  stage:                           lint
+  <<:                              *docker-env
+  script:
+    # add clippy is temporary until tests in substrate are not fixed
+    # https://github.com/paritytech/substrate/pull/9637
+    - rustup component add clippy --toolchain nightly-x86_64-unknown-linux-gnu
+    - cargo +nightly clippy --all-targets -- -D warnings
+
+#### stage:                       test
+
+deny:
+  stage:                           test
+  <<:                              *docker-env
+  <<:                              *collect-artifacts
+  script:
+    - cargo deny check --hide-inclusion-graph
+  after_script:
+    - mkdir -p ./artifacts
+    - echo "___Complete logs can be found in the artifacts___"
+    - cargo deny check --hide-inclusion-graph 2> cargo_deny.log
+
+test:
+  stage:                           test
+  <<:                              *docker-env
+  script:
+    - cargo test --all  --verbose
+
+check:
+  stage:                           test
+  <<:                              *docker-env
+  script:
+    - cargo check --all  --verbose
```

Comment:
Do I get this right that `cargo check` will be performed after `cargo test`? I don't think it makes much sense - `check` is more lightweight (not doing compilation, just static analysis) and I think there isn't anything that would slip through `test` run.

Reply:
Not exactly: jobs `deny`, `test` and `check` are executed on the stage `test` which means that they run simultaneously on different runners. But I can divide it to different stages, just not sure which way is the best.

Reply:
Running in parallel sounds good, we can keep it as-is.